stdlib/src/collections/deque.mojo

```diff
@@ -0,0 +1,313 @@
+# ===----------------------------------------------------------------------=== #
+# Copyright (c) 2024, Modular Inc. All rights reserved.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions:
+# https://llvm.org/LICENSE.txt
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+"""Defines the Deque type.
+
+You can import these APIs from the `collections` package.
+
+Examples:
+
+```mojo
+from collections import Deque
+```
+"""
+
+from bit import bit_ceil
+from collections import Optional
+from memory import UnsafePointer
+
+from builtin._documentation import doc_private
+
+# ===----------------------------------------------------------------------===#
+# Deque
+# ===----------------------------------------------------------------------===#
+
+
+struct Deque[ElementType: CollectionElement](
+    Movable, ExplicitlyCopyable, Sized, Boolable
+):
+    """Implements a double-ended queue.
+
+    It supports pushing and popping from both ends in O(1) time resizing the
+    underlying storage as needed.
+
+    Parameters:
+        ElementType: The type of the elements in the deque.
+            Must implement the trait `CollectionElement`.
+    """
+
+    # ===-------------------------------------------------------------------===#
+    # Aliases
+    # ===-------------------------------------------------------------------===#
+
+    alias default_capacity: Int = 64
+    """The default capacity of the deque: must be the power of 2."""
+
+    # ===-------------------------------------------------------------------===#
+    # Fields
+    # ===-------------------------------------------------------------------===#
+
+    var data: UnsafePointer[ElementType]
+    """The underlying storage for the deque."""
+
+    var head: Int
+    """The index of the head: points the first element of the deque."""
+
+    var tail: Int
+    """The index of the tail: points behind the last element of the deque."""
+
+    var capacity: Int
+    """The amount of elements that can fit in the deque without resizing it."""
+
+    var minlen: Int
```

Comment:
Correct, it has nothing to do with the minimum number of elements in the deque, so I will rename it.
To give some more information on this, my implementation is based on the buffers with the capacities of powers of 2 (mainly for efficiency of physical index calculations using `&` instead of `%`). So, if the user gives the `min_capacity` arg of 100, the real min capacity of the deque will be 128.
This works a bit different with `maxlen`: if you ask for maxlen of 100, you will never have more than 100 elements in the deque, but the supporting buffer size will be 128 anyway.